subprojects/core/src/main/java/org/gradle/api/internal/changedetection/rules/InputPropertyValueTaskStateChanges.java

```diff
@@ -16,40 +16,52 @@
 
 package org.gradle.api.internal.changedetection.rules;
 
+import com.google.common.collect.ImmutableSet;
 import com.google.common.collect.ImmutableSortedMap;
 import org.gradle.api.internal.TaskInternal;
+import org.gradle.api.internal.changedetection.state.ImplementationSnapshot;
 import org.gradle.api.internal.changedetection.state.TaskExecution;
 import org.gradle.api.internal.changedetection.state.ValueSnapshot;
 
-import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 
 class InputPropertyValueTaskStateChanges extends SimpleTaskStateChanges {
     private final TaskInternal task;
-    private final Set<String> changed;
+    private final Set<String> valueChanges;
+    private final Set<String> implementationChanges;
 
     public InputPropertyValueTaskStateChanges(TaskExecution previousExecution, TaskExecution currentExecution, TaskInternal task) {
         ImmutableSortedMap<String, ValueSnapshot> previousInputProperties = previousExecution.getInputProperties();
-        changed = new HashSet<String>();
+        ImmutableSet.Builder<String> valueChangesBuilder = ImmutableSet.builder();
+        ImmutableSet.Builder<String> implementationChangesBuilder = ImmutableSet.builder();
```

Comment:
I would have used a `Map` with a value of `"Implementation"` vs `"Value"` to save a bit of memory, but I guess this is cool, too.

BTW, if we don't us a `Map`, I think using a `List` instead of a `Set` here would work better, too, as it would require no uniqueness checks when we build the collection.